### ISSUE TITLE
Re-entrance issues with Forge's EventBus

### ISSUE THREAD

**Issue by Johni0702** [SpongePowered/Mixin#369]
Forge: 1.14.4-28.1.70
Mixin: 0.8-preview (10af60b)
Used to work before bytecode loading was moved to ModLauncher (using Forge 1.14.4-28.0.75 and Mixin a158972).

Forge's EventBus agressively loads the parent class of any class it handles ([source](https://github.com/MinecraftForge/EventBus/blob/3f83a9efdc1468a9f3b3f93fa3cb3471fbf7d653/src/main/java/net/minecraftforge/eventbus/EventSubclassTransformer.java#L55-L68) explicitly claims that this is legal) which quickly results in re-entrance errors.

(Pardon the fact that the following source files are in their 1.12 version. At build time they're run through a pre-processor which works as you'd expect and also automatically remaps any class/method/field references between versions as far as possible.)

E.g. I have a [`MixinIntegratedServer`](https://github.com/Johni0702/BetterPortals/blob/5b03536f197df1e3fc48c4fc2c993fe2a70cf9dd/src/view/java/de/johni0702/minecraft/view/impl/mixin/MixinIntegratedServer.java#L38) which extends `MinecraftServer` (no longer strictly necessary but makes the bug simpler than the second example) and a [`MixinThreadTaskExecutor`](https://github.com/Johni0702/BetterPortals/blob/5b03536f197df1e3fc48c4fc2c993fe2a70cf9dd/src/view/java/de/johni0702/minecraft/view/impl/mixin/MixinThreadTaskExecutor.java#L26) (`ThreadTaskExecutor` is a super-class of `MinecraftServer`).
When mixin tries to prepare the `MixinIntegratedServer`, it asks ModLauncher for its bytecode which causes the EventBus transformer to recursively load `MinecraftServer` which in turn loads its super-class `ThreadTaskExecutor`. In the process of loading `ThreadTaskExecutor`, ModLauncher asks Mixin to transform that class which then detects the re-entrance and bails out.
<details>

```
[11:31:00] [main/WARN] [mixin/]: Re-entrance detected during prepare phase, this will cause serious problems.
org.spongepowered.asm.mixin.transformer.throwables.ReEntrantTransformerError: Re-entrance error.
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:266) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:206) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinTransformationHandler.processClass(MixinTransformationHandler.java:109) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.launch.MixinLaunchPlugin.processClass(MixinLaunchPlugin.java:131) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at cpw.mods.modlauncher.LaunchPluginHandler.offerClassNodeToPlugins(LaunchPluginHandler.java:80) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.ClassTransformer.transform(ClassTransformer.java:115) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader$DelegatedClassLoader.findClass(TransformingClassLoader.java:241) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader.loadClass(TransformingClassLoader.java:128) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader.loadClass(TransformingClassLoader.java:98) ~[modlauncher-4.1.0.jar:?] {}
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[?:1.8.0_222] {}
	at java.lang.ClassLoader.defineClass1(Native Method) ~[?:1.8.0_222] {}
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763) ~[?:1.8.0_222] {}
	at java.lang.ClassLoader.defineClass(ClassLoader.java:642) ~[?:1.8.0_222] {}
	at cpw.mods.modlauncher.TransformingClassLoader.loadClass(TransformingClassLoader.java:129) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader.loadClass(TransformingClassLoader.java:98) ~[modlauncher-4.1.0.jar:?] {}
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[?:1.8.0_222] {}
	at java.lang.ClassLoader.defineClass1(Native Method) ~[?:1.8.0_222] {}
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763) ~[?:1.8.0_222] {}
	at java.lang.ClassLoader.defineClass(ClassLoader.java:642) ~[?:1.8.0_222] {}
	at cpw.mods.modlauncher.TransformingClassLoader.loadClass(TransformingClassLoader.java:129) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader.loadClass(TransformingClassLoader.java:98) ~[modlauncher-4.1.0.jar:?] {}
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[?:1.8.0_222] {}
	at net.minecraftforge.eventbus.EventSubclassTransformer.buildEvents(EventSubclassTransformer.java:62) ~[eventbus-1.0.0-service.jar:?] {}
	at net.minecraftforge.eventbus.EventSubclassTransformer.transform(EventSubclassTransformer.java:44) ~[eventbus-1.0.0-service.jar:?] {}
	at net.minecraftforge.eventbus.EventBusEngine.processClass(EventBusEngine.java:20) ~[eventbus-1.0.0-service.jar:?] {}
	at net.minecraftforge.eventbus.service.ModLauncherService.processClass(ModLauncherService.java:19) ~[eventbus-1.0.0-service.jar:1.0.0+54+3f83a9e] {}
	at cpw.mods.modlauncher.serviceapi.ILaunchPluginService.processClass(ILaunchPluginService.java:102) ~[modlauncher-4.1.0.jar:4.1.0+62+5bfa59b] {}
	at cpw.mods.modlauncher.LaunchPluginHandler.offerClassNodeToPlugins(LaunchPluginHandler.java:80) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.ClassTransformer.transform(ClassTransformer.java:115) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader$DelegatedClassLoader.findClass(TransformingClassLoader.java:241) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader.buildTransformedClassNodeFor(TransformingClassLoader.java:133) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.LaunchPluginHandler.lambda$null$9(LaunchPluginHandler.java:90) ~[modlauncher-4.1.0.jar:?] {}
	at org.spongepowered.asm.launch.MixinLaunchPlugin.getClassNode(MixinLaunchPlugin.java:196) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinInfo.loadMixinClass(MixinInfo.java:1259) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinInfo.<init>(MixinInfo.java:812) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinConfig.prepareMixins(MixinConfig.java:711) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinConfig.prepare(MixinConfig.java:648) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.prepareConfigs(MixinProcessor.java:474) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.select(MixinProcessor.java:399) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.checkSelect(MixinProcessor.java:375) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:273) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:206) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinTransformationHandler.processClass(MixinTransformationHandler.java:109) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.launch.MixinLaunchPlugin.processClass(MixinLaunchPlugin.java:131) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at cpw.mods.modlauncher.LaunchPluginHandler.offerClassNodeToPlugins(LaunchPluginHandler.java:80) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.ClassTransformer.transform(ClassTransformer.java:115) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader$DelegatedClassLoader.findClass(TransformingClassLoader.java:241) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader.loadClass(TransformingClassLoader.java:128) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader.loadClass(TransformingClassLoader.java:98) ~[modlauncher-4.1.0.jar:?] {}
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[?:1.8.0_222] {}
	at java.lang.Class.forName0(Native Method) ~[?:1.8.0_222] {}
	at java.lang.Class.forName(Class.java:348) ~[?:1.8.0_222] {}
	at net.minecraftforge.userdev.FMLUserdevClientLaunchProvider.lambda$launchService$0(FMLUserdevClientLaunchProvider.java:55) ~[forge-1.14.4-28.1.70_mapped_snapshot_20191009-1.14.3-recomp.jar:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandlerDecorator.launch(LaunchServiceHandlerDecorator.java:37) [modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:54) [modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:72) [modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.Launcher.run(Launcher.java:81) [modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.Launcher.main(Launcher.java:65) [modlauncher-4.1.0.jar:?] {}
	at net.minecraftforge.userdev.LaunchTesting.main(LaunchTesting.java:101) [forge-1.14.4-28.1.70_mapped_snapshot_20191009-1.14.3-recomp.jar:?] {}
```
</details>

Another, less direct example (also, the exception is throw from a different location):
[`MixinServerWorld`](https://github.com/Johni0702/BetterPortals/blob/5b03536f197df1e3fc48c4fc2c993fe2a70cf9dd/src/view/java/de/johni0702/minecraft/view/impl/mixin/MixinServerWorldEventHandler.java#L30) (name in source is still based on 1.12) which [references](https://github.com/Johni0702/BetterPortals/blob/5b03536f197df1e3fc48c4fc2c993fe2a70cf9dd/src/view/java/de/johni0702/minecraft/view/impl/mixin/MixinServerWorldEventHandler.java#L54) [`ViewEntity`](https://github.com/Johni0702/BetterPortals/blob/5b03536f197df1e3fc48c4fc2c993fe2a70cf9dd/src/view/kotlin/de/johni0702/minecraft/view/impl/server/ViewEntity.kt#L39) (so Mixin fetches its bytecode) which [extends](https://github.com/Johni0702/BetterPortals/blob/5b03536f197df1e3fc48c4fc2c993fe2a70cf9dd/src/view/kotlin/de/johni0702/minecraft/view/impl/server/ViewEntity.kt#L44) `ServerPlayerEntity` (again, source is 1.12.2 and automatically remapped on build, the 1.14 version does say `ServerPlayerEntity`) (so EventBus loads it) which finally results in Mixin bailing out again because there are other mixins for ServerPlayerEntity (e.g. [this one](https://github.com/Johni0702/BetterPortals/blob/5b03536f197df1e3fc48c4fc2c993fe2a70cf9dd/src/transition/java/de/johni0702/minecraft/betterportals/impl/transition/mixin/MixinPlayerList.java#L20)).
<details>

```
[12:00:26] [main/WARN] [mixin/]: Re-entrance detected, this will cause serious problems.
org.spongepowered.asm.mixin.transformer.throwables.ReEntrantTransformerError: Re-entrance error.
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:319) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:206) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinTransformationHandler.processClass(MixinTransformationHandler.java:109) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.launch.MixinLaunchPlugin.processClass(MixinLaunchPlugin.java:131) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at cpw.mods.modlauncher.LaunchPluginHandler.offerClassNodeToPlugins(LaunchPluginHandler.java:80) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.ClassTransformer.transform(ClassTransformer.java:115) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader$DelegatedClassLoader.findClass(TransformingClassLoader.java:241) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader.loadClass(TransformingClassLoader.java:128) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader.loadClass(TransformingClassLoader.java:98) ~[modlauncher-4.1.0.jar:?] {}
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[?:1.8.0_222] {}
	at net.minecraftforge.eventbus.EventSubclassTransformer.buildEvents(EventSubclassTransformer.java:62) ~[eventbus-1.0.0-service.jar:?] {}
	at net.minecraftforge.eventbus.EventSubclassTransformer.transform(EventSubclassTransformer.java:44) ~[eventbus-1.0.0-service.jar:?] {}
	at net.minecraftforge.eventbus.EventBusEngine.processClass(EventBusEngine.java:20) ~[eventbus-1.0.0-service.jar:?] {}
	at net.minecraftforge.eventbus.service.ModLauncherService.processClass(ModLauncherService.java:19) ~[eventbus-1.0.0-service.jar:1.0.0+54+3f83a9e] {}
	at cpw.mods.modlauncher.serviceapi.ILaunchPluginService.processClass(ILaunchPluginService.java:102) ~[modlauncher-4.1.0.jar:4.1.0+62+5bfa59b] {}
	at cpw.mods.modlauncher.LaunchPluginHandler.offerClassNodeToPlugins(LaunchPluginHandler.java:80) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.ClassTransformer.transform(ClassTransformer.java:115) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader$DelegatedClassLoader.findClass(TransformingClassLoader.java:241) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader.buildTransformedClassNodeFor(TransformingClassLoader.java:133) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.LaunchPluginHandler.lambda$null$9(LaunchPluginHandler.java:90) ~[modlauncher-4.1.0.jar:?] {}
	at org.spongepowered.asm.launch.MixinLaunchPlugin.getClassNode(MixinLaunchPlugin.java:196) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.launch.MixinLaunchPlugin.getClassNode(MixinLaunchPlugin.java:184) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.ClassInfo.forName(ClassInfo.java:1807) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.ClassInfo.forType(ClassInfo.java:1861) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.ClassInfo.forDescriptor(ClassInfo.java:1840) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.transformField(MixinPreProcessorStandard.java:750) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.transform(MixinPreProcessorStandard.java:722) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.attach(MixinPreProcessorStandard.java:296) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinPreProcessorStandard.createContextFor(MixinPreProcessorStandard.java:266) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinInfo.createContextFor(MixinInfo.java:1236) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.apply(MixinApplicatorStandard.java:278) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.applyMixins(TargetClassContext.java:345) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:529) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:334) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:206) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.mixin.transformer.MixinTransformationHandler.processClass(MixinTransformationHandler.java:109) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at org.spongepowered.asm.launch.MixinLaunchPlugin.processClass(MixinLaunchPlugin.java:131) ~[mixin-0.8-preview-SNAPSHOT.jar:0.8-preview-SNAPSHOT+unknown-b0.git-unknown] {}
	at cpw.mods.modlauncher.LaunchPluginHandler.offerClassNodeToPlugins(LaunchPluginHandler.java:80) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.ClassTransformer.transform(ClassTransformer.java:115) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader$DelegatedClassLoader.findClass(TransformingClassLoader.java:241) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader.loadClass(TransformingClassLoader.java:128) ~[modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.TransformingClassLoader.loadClass(TransformingClassLoader.java:98) ~[modlauncher-4.1.0.jar:?] {}
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[?:1.8.0_222] {}
	at net.minecraft.client.main.Main.main(Main.java:107) ~[forge-1.14.4-28.1.70_mapped_snapshot_20191009-1.14.3-recomp.jar:?] {re:classloading,pl:runtimedistcleaner:A}
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_222] {}
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_222] {}
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_222] {}
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_222] {}
	at net.minecraftforge.userdev.FMLUserdevClientLaunchProvider.lambda$launchService$0(FMLUserdevClientLaunchProvider.java:55) ~[forge-1.14.4-28.1.70_mapped_snapshot_20191009-1.14.3-recomp.jar:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandlerDecorator.launch(LaunchServiceHandlerDecorator.java:37) [modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:54) [modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:72) [modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.Launcher.run(Launcher.java:81) [modlauncher-4.1.0.jar:?] {}
	at cpw.mods.modlauncher.Launcher.main(Launcher.java:65) [modlauncher-4.1.0.jar:?] {}
	at net.minecraftforge.userdev.LaunchTesting.main(LaunchTesting.java:101) [forge-1.14.4-28.1.70_mapped_snapshot_20191009-1.14.3-recomp.jar:?] {}
```
</details>

**Comment by Mumfrey** [comment]
I don't have a good answer for this, I'm going to need to have a chat with cpw I think and try and work out the best way to deal with this situation. The problem is that there's no "general solution" for dealing with re-entrance here.

The comments you linked are in general true of the transformer contract, it **is** right to assert that being re-entrant is ok for transformers, however not in this side-loading context. In launchwrapper I handled this with delegation and exclusion of re-entrant transformers, but a more desirable solution in this situation is that the contract of transformers is clarified.

In an ideal world, transformer is okay to be re-entrant when classloading, and not for side-loading. Since `reason` is now promulgated through class loading and transformation it would make sense for transformers to adopt this approach.

On the other hand, it looks like the transformer in question is only re-entrant in order to determine hierarchy, something which is already handled more elegantly [here](https://github.com/cpw/modlauncher/blob/master/src/main/java/cpw/mods/modlauncher/TransformerClassWriter.java#L77) and so the problem is being caused by the precise thing that `ClassInfo` is designed to fix, viz. class metadata.

I keep trying to think through how to deal with re-entrance, and I can think of ways to deal with the larger problem of re-entrance for classloading, but not for side-loading. Since ultimately if re-entrance happens during classload it's actually safe - if stupidly inefficient - to tear down the whole state and start again. The main issue we run into there, is that if the re-entrant transformer is idempotent then you get stuck in a loop.

_Which goes back to needing a clarification of contract. If the contract of transformers is such that "it's ok to be re-entrant when reason is `classloading` and **not ok** to be re-entrant for plugin reasons._ I really need @cpw to weigh in on this.

If that's the case it can be fixed in one of two places. Either modlauncher itself can provide some kind of re-entrant locking whereby `classloading` can only be entered if a plugin context is **not** active. Alternatively the issue can be fixed in the EventSubclassTransformer by having it skip the transformation pass when the reason is not `classloading`. The benefit of the former is that it encodes an implicit contract of no-entrance into a more formal one. If modlauncher throws a specific exception when trying to enter `classloading` illegally, it can also catch that in order to identify and elide re-entrant transformers in much the same way I did with the delegation list when running under LaunchWrapper.

Again, this is a little bit beyond my capability to fix alone. I don't consider mixin's behaviour inappropriate here because there is **so much state** that needs to be generated to do the work efficiently that re-entrance just torpedoes. At least in the short-term I'm not considering this a bug since I think the correct approach is to fix this issue co-operatively. I will try and find cpw to have a chat with him soon and see what his opinion is.

Thank you for the detailed report. I have removed the parenthetical from your issue title and replaced it with the appropriate tag.
